src/saveload/gamelog_sl.cpp

```diff
@@ -15,147 +15,317 @@
+	{
+		if (lc->ct != GLCT_GRFCOMPAT) return;
+		SlObject(lc, this->GetDescription());
+	}
+
+	void Save(LoggedChange *lc) const override { this->GenericSaveLoad(lc); }
+	void Load(LoggedChange *lc) const override { this->GenericSaveLoad(lc); }
+	void LoadCheck(LoggedChange *lc) const override { this->GenericSaveLoad(lc); }
 };
 
-static const SaveLoad _glog_grfbug_desc[] = {
-	SLE_VAR(LoggedChange, grfbug.data,       SLE_UINT64),
-	SLE_VAR(LoggedChange, grfbug.grfid,      SLE_UINT32),
-	SLE_VAR(LoggedChange, grfbug.bug,        SLE_UINT8),
+class SlGamelogGrfparam : public DefaultSaveLoadHandler<SlGamelogGrfparam, LoggedChange> {
+public:
+	inline static const SaveLoad description[] = {
+		SLE_VAR(LoggedChange, grfparam.grfid,    SLE_UINT32),
+	};
+
+	void GenericSaveLoad(LoggedChange *lc) const
+	{
+		if (lc->ct != GLCT_GRFPARAM) return;
+		SlObject(lc, this->GetDescription());
+	}
+
+	void Save(LoggedChange *lc) const override { this->GenericSaveLoad(lc); }
+	void Load(LoggedChange *lc) const override { this->GenericSaveLoad(lc); }
+	void LoadCheck(LoggedChange *lc) const override { this->GenericSaveLoad(lc); }
 };
 
-static const SaveLoad _glog_emergency_desc[] = {
-	SLE_CONDNULL(0, SL_MIN_VERSION, SL_MIN_VERSION), // Just an empty list, to keep the rest of the code easier.
+class SlGamelogGrfmove : public DefaultSaveLoadHandler<SlGamelogGrfmove, LoggedChange> {
+public:
+	inline static const SaveLoad description[] = {
+		SLE_VAR(LoggedChange, grfmove.grfid,     SLE_UINT32),
+		SLE_VAR(LoggedChange, grfmove.offset,    SLE_INT32),
+	};
+
+	void GenericSaveLoad(LoggedChange *lc) const
+	{
+		if (lc->ct != GLCT_GRFMOVE) return;
+		SlObject(lc, this->GetDescription());
+	}
+
+	void Save(LoggedChange *lc) const override { this->GenericSaveLoad(lc); }
+	void Load(LoggedChange *lc) const override { this->GenericSaveLoad(lc); }
+	void LoadCheck(LoggedChange *lc) const override { this->GenericSaveLoad(lc); }
 };
 
-static const SaveLoadTable _glog_desc[] = {
-	_glog_mode_desc,
-	_glog_revision_desc,
-	_glog_oldver_desc,
-	_glog_setting_desc,
-	_glog_grfadd_desc,
-	_glog_grfrem_desc,
-	_glog_grfcompat_desc,
-	_glog_grfparam_desc,
-	_glog_grfmove_desc,
-	_glog_grfbug_desc,
-	_glog_emergency_desc,
+class SlGamelogGrfbug : public DefaultSaveLoadHandler<SlGamelogGrfbug, LoggedChange> {
+public:
+	inline static const SaveLoad description[] = {
+		SLE_VAR(LoggedChange, grfbug.data,       SLE_UINT64),
+		SLE_VAR(LoggedChange, grfbug.grfid,      SLE_UINT32),
+		SLE_VAR(LoggedChange, grfbug.bug,        SLE_UINT8),
+	};
+
+	void GenericSaveLoad(LoggedChange *lc) const
+	{
+		if (lc->ct != GLCT_GRFBUG) return;
+		SlObject(lc, this->GetDescription());
+	}
+
+	void Save(LoggedChange *lc) const override { this->GenericSaveLoad(lc); }
+	void Load(LoggedChange *lc) const override { this->GenericSaveLoad(lc); }
+	void LoadCheck(LoggedChange *lc) const override { this->GenericSaveLoad(lc); }
 };
 
-static_assert(lengthof(_glog_desc) == GLCT_END);
```

Comment:
Would it make sense to add something similar to `SlGamelogAction::description` but with a `+ 1`? Alternatively you could say that if someone adds a game log action but do not check it survives a save-load, then the game log action is pointless and should not be added.

Reply:
Exactly the last part. The assert was there as the length of the list was important. That is no longer the case. And especially as you can now view a savegame, I think we have to assume the developer validates it :D